# Deduper Psuedocode

## Define the problem
The goal of this algorithm is to eliminate PCR duplicates from a SAM file of mapped reads. The input SAM file should be sorted first by chromosome and second by start position. We will define PCR duplicates as reads with the same 5' start position on the same strand and chromosome which also possess the same unique molecular identifier (UMI). However, the 5' start position may need to be adjusted if one read contains soft clipping; this can be identified using the CIGAR string. If PCR duplicates are identified, this algorithm will need to remove all but the first read (i.e. if there are 6 duplicates, only the first read is retained and the other 5 are thrown out). Additionally, this algorithm needs to avoid reading large chunks of the file into memory at once, as this would be computationally inefficent. 

## Input/Output Examples

## High Level Functions
```python

def extract_SAM_info(line: str) -> dict:
    '''Given an alignment line of a SAM file, extracts the UMI, CHR, POS, CIGAR, and STRAND and returns them in the form of a dictionary.'''
    Input: "NS500451:154:HWKTMBGXX:1:11101:17732:1263:CGCCTATT	0	2	30402279	36	71M	*	0	0	GGCTCCTGCTGTGAGAAGGGATGGGCTGGACTGGCTCTCTTCTCCATGGTTCATGGAGTGGATGATACATT	6<EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE	MD:Z:71	NH:i:1	HI:i:1	NM:i:0	SM:i:36	XQ:i:40	X2:i:0	XO:Z:UU"
    Output: {"UMI": "CGCCTATT", "CHR" : 2, "POS":304022, "CIGAR":"71M", "STRAND":"+"}

def soft_clip_adj(SAM_info: dict) -> dict:
    '''Extracts the position and cigar string from a dictionary of values generated by extract_SAM_info. If soft clipping occurs, it returns the dictionary with the position adjusted for softclipping. If it is a negative strand strand, it needs to look for soft clipping at the end of the cigar string and also account for deletions. If it is a positive strand, it needs to look for soft clipping at the beginning of the cigar string'''
    Input: {"UMI": "CGCCTATT", "CHR" : 2, "POS":43, "CIGAR":"3S12M", "STRAND":"+"}
	Expected output: {"UMI": "CGCCTATT", "CHR" : 2, "POS":40, "CIGAR":"3S12M", "STRAND":"+"}

def filter_pcr_dupes(UMI_list: list) -> list:
    '''Given a list of format [[line, (POS, STRAND)]] comprised of SAM file entries with the same UMI and same CHR, return a list of lines with PCR duplicates filtered out, selecting the first entry of any duplicates'''
     Input: [["NS500451:154:HWKTMBGXX:1:11101:89435:58226:TCGTAGGT	0	2	76708822	36	71M	*	0	0	CCTGCTCATGGATACTCTTGAAGGATTGTCCAGTAAACTGGAAGCTGGTTTTGGAAGTTCCTCCTTCACCA	6AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE	MD:Z:71	NH:i:1	HI:i:1	NM:i:0	SM:i:36	XQ:i:40	X2:i:0	XO:Z:UU",
     (76708822, "+")], ["NS500451:154:HWKTMBGXX:1:11101:19377:1220:TCGTAGGT	0	2	76708822	36	71M	*	0	0	CCTGCTCATGGATACTCTTGAAGGATTGTCCAGTAAACTGGAAGCTGGTTTTGGAAGTTCCTCCTTCACCA	6AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE	MD:Z:71	NH:i:1	HI:i:1	NM:i:0	SM:i:36	XQ:i:40	X2:i:0	XO:Z:UU",
     (76708822, "+")], ["NS500451:154:HWKTMBGXX:1:11101:11995:1145:AGCTACCA	0	2	76710746	36	71M	*	0	0	TGCATAACTCGTGCTGGTTTCCTCCTTTGTGGGGACGTGATAGGTCGAGTACCTGAAGTCTCTTCTTCTGT	6<EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE	MD:Z:71	NH:i:1	HI:i:1	NM:i:0	SM:i:36	XQ:i:40	X2:i:0	XO:Z:UU",
     (76710746, "+")]]
     Output: ["NS500451:154:HWKTMBGXX:1:11101:89435:58226:TCGTAGGT	0	2	76708822	36	71M	*	0	0	CCTGCTCATGGATACTCTTGAAGGATTGTCCAGTAAACTGGAAGCTGGTTTTGGAAGTTCCTCCTTCACCA	6AEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE	MD:Z:71	NH:i:1	HI:i:1	NM:i:0	SM:i:36	XQ:i:40	X2:i:0	XO:Z:UU",
     "NS500451:154:HWKTMBGXX:1:11101:11995:1145:AGCTACCA	0	2	76710746	36	71M	*	0	0	TGCATAACTCGTGCTGGTTTCCTCCTTTGTGGGGACGTGATAGGTCGAGTACCTGAAGTCTCTTCTTCTGT	6<EEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE	MD:Z:71	NH:i:1	HI:i:1	NM:i:0	SM:i:36	XQ:i:40	X2:i:0	XO:Z:UU"]
```

## Psuedocode

```
Create list of 96 UMIs from STL96.txt called UMI_list

For umi in UMI_list:
    append reverse complement to UMI_list

With open SAM file and open output file:
    chr_umi_list = []
    if line starts with @:
        print to output file

    
    for UMI in UMI_list:
        count = 0 
        for line in SAM file:
        
            if line does not start with @ and count == 0:
                extract CHR
                current_chr = CHR
                extract umi
                if umi == UMI:
                    info = extract_SAM_info(line)
                    info = soft_clip_adj(info)
                    append [line, (info["POSITION"], info["STRAND"])] to chr_umi_list

            else if count != 0:
                extract CHR
                if CHR == current_chr:
                    extract umi
                    if umi == UMI:
                        info = extract_SAM_info(line)
                        info = soft_clip_adj(info)
                        append [line, (info["POSITION"], info["STRAND"])] to chr_umi_list

                if CHR != current_chr:
                    chr_umi_list = filter_pcr_dupes(chr_umi_list)
                    for entry in chr_umi_list:
                        print line to output file
                    clear chr_umi_list
                    current_chr = CHR
                    extract umi
                    if umi == UMI
                    info = extract_SAM_info(line)
                        info = soft_clip_adj(info)
                        append [line, (info["POSITION"], info["STRAND"])] to chr_umi_list
            count += 1


```